### PR TITLE
New AWS module for managing RDS instance using newer boto3 API

### PR DESCRIPTION
**Pull request by aioue**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

rds_v2
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel e396d5d508) last updated 2016/08/24 12:13:10 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD 368ca738fa) last updated 2016/08/24 12:13:21 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD 0749ce6faa) last updated 2016/08/24 12:13:30 (GMT +1000)
```
##### SUMMARY

This module manages the creation, modification, deletion, snapshot, restore, reboot, replicate and promote of RDS Instances in AWS.

Note: Technically this is a rewrite of the existing core RDS module and while I believe rewrites using older API's aren't being included until 2.3 I thought people could still use this in the meantime, especially if support for Aurora is required. 

```
TASK [print new RDS instance] **************************************************
ok: [localhost] => {
    "msg": {
        "changed": true, 
        "rds": {
            "DBInstances": [
                {
                    "AllocatedStorage": 10, 
                    "AutoMinorVersionUpgrade": true, 
                    "AvailabilityZone": "ap-southeast-2a", 
                    "BackupRetentionPeriod": 10, 
                    "CACertificateIdentifier": "rds-ca-2015", 
                    "CopyTagsToSnapshot": false, 
                    "DBInstanceClass": "db.t2.medium", 
                    "DBInstanceIdentifier": "test-instance", 
                    "DBInstanceStatus": "available", 
                    "DBParameterGroups": [
                        {
                            "DBParameterGroupName": "test-param-group", 
                            "ParameterApplyStatus": "in-sync"
                        }
                    ], 
                    "DBSecurityGroups": [], 
                    "DBSubnetGroup": {
                        "DBSubnetGroupDescription": "test subnet group", 
                        "DBSubnetGroupName": "test-subnet-group", 
                        "SubnetGroupStatus": "Complete", 
                        "Subnets": [
                            {
                                "SubnetAvailabilityZone": {
                                    "Name": "ap-southeast-2a"
                                }, 
                                "SubnetIdentifier": "subnet-75ed1123", 
                                "SubnetStatus": "Active"
                            }, 
                            {
                                "SubnetAvailabilityZone": {
                                    "Name": "ap-southeast-2b"
                                }, 
                                "SubnetIdentifier": "subnet-6ae33123", 
                                "SubnetStatus": "Active"
                            }
                        ], 
                        "VpcId": "vpc-aad24acf"
                    }, 
                    "DbInstancePort": 0, 
                    "DbiResourceId": "db-7ZINBVLFLHNOT5ROFK76T4XXXX", 
                    "Endpoint": {
                        "Address": "test-instance2.c4nb0jta2bug.ap-southeast-2.rds.amazonaws.com", 
                        "HostedZoneId": "Z32T0VRHXEXXXX", 
                        "Port": 3306
                    }, 
                    "Engine": "mysql", 
                    "EngineVersion": "5.6.27", 
                    "InstanceCreateTime": "2016-09-03T07:36:28.832000+00:00", 
                    "LatestRestorableTime": "2016-09-03T07:40:00+00:00", 
                    "LicenseModel": "general-public-license", 
                    "MasterUsername": "testuser", 
                    "MonitoringInterval": 0, 
                    "MultiAZ": false, 
                    "OptionGroupMemberships": [
                        {
                            "OptionGroupName": "test-option-group", 
                            "Status": "in-sync"
                        }
                    ], 
                    "PendingModifiedValues": {}, 
                    "PreferredBackupWindow": "17:54-18:24", 
                    "PreferredMaintenanceWindow": "fri:12:54-fri:13:24", 
                    "PubliclyAccessible": false, 
                    "ReadReplicaDBInstanceIdentifiers": [], 
                    "StorageEncrypted": false, 
                    "StorageType": "standard", 
                    "VpcSecurityGroups": [
                        {
                            "Status": "active", 
                            "VpcSecurityGroupId": "sg-d188c123"
                        }
                    ]
                }
            ], 
            "ResponseMetadata": {
                "HTTPStatusCode": 200, 
                "RequestId": "022114a4-71aa-11e6-8a7d-bd9007adb1d5"
            }
        }
    }
}
```

### CODE REVIEW

**Comment by aioue** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2845 by aioue (not original author)

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 1]
```diff
@@ -0,0 +1,1148 @@
...
+        engine=dict(type='str', choices=['mariadb', 'MySQL', 'oracle-se1', 'oracle-se', 'oracle-ee', 'sqlserver-ee', 'sqlserver-se', 'sqlserver-ex', 'sqlserver-web', 'postgres', 'aurora'], required=False),
+        master_username=dict(type='str', required=False),
+        master_user_password=dict(type='str', no_log=True, required=False),
+        vpc_security_group_ids=dict(type='list', required=False),
+        apply_immediately=dict(type='bool', default=False),
+        availability_zone=dict(type='str', required=False),
+        db_subnet_group_name=dict(type='str', required=False),
+        preferred_maintenance_window=dict(type='str', required=False),
+        db_parameter_group_name=dict(type='str', required=False),
+        backup_retention_period=dict(type=int, required=False),
```
Comment (aioue) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:1087]: Should be `type='int'`.

**Comment by aioue** [comment]
After that minor change on line 1087, I was able to successfully use this module to create a storage encrypted postrgesql DB. @naslanidis

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 2]
```diff
@@ -0,0 +1,1148 @@
...
+    description:
+      - Master database username. Used only when command=create.
+    required: false
+    default: null
+  master_user_password:
+    description:
+      - Password for the master database username. Used only when command=create modify.
+    required: false
+    default: null
+  region:
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:69]: Not required - added by doc fragement

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 3]
```diff
@@ -0,0 +1,1148 @@
...
+        params['DBInstanceIdentifier'] = instance_name
+    else:
+        params['DBInstanceIdentifier'] = module.params.get('db_instance_identifier')
+
+    try:
+        response = client.describe_db_instances(**params)
+
+    except botocore.exceptions.ClientError as e:
+        if 'DBInstanceNotFound' not in e.message:
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:679]: Handle boto3 exceptions as per the guidelines.

E.g.
```module.fail_json(msg=e.message, **camel_dict_to_snake_dict(e.response))```

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 4]
```diff
@@ -0,0 +1,1148 @@
...
+    if db_instance:
+        response = db_instance
+        changed=False
+    else:
+        try:
+            response = client.create_db_instance(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:734]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 5]
```diff
@@ -0,0 +1,1148 @@
...
+
+        if module.params.get('wait'):
+            status_achieved, response = wait_for_status(client, module, 'available')
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:737]: There's a new AWS Retry generic handler that you can use here.  Check out the ec2 shared code in module_utils

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 6]
```diff
@@ -0,0 +1,1148 @@
...
+        params['DBPortNumber'] = module.params.get('port')
+
+    # Get current instance so we can see if anything was actually modified and set changed accordingly
+    before_modify_instance = get_db_instance(client, module)
+
+    try:
+        response = client.modify_db_instance(**params)
+    except botocore.exceptions.ClientError as e:
+        e = get_exception()
+        module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:768]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 7]
```diff
@@ -0,0 +1,1148 @@
...
+    if db_snapshot['DBSnapshots']:
+        response = db_snapshot
+        changed=False
+    else:
+        try:
+            response = client.create_db_snapshot(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:820]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 8]
```diff
@@ -0,0 +1,1148 @@
...
+        if not db_snapshot:
+            response = None
+            changed=False
+        else:
+            try:
+                response = client.delete_db_snapshot(**params)
+                changed = True
+            except botocore.exceptions.ClientError as e:
+                e = get_exception()
+                module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:854]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 9]
```diff
@@ -0,0 +1,1148 @@
...
+        if not db_instance:
+            response = None
+            changed=False
+        else:
+            try:
+                response = client.delete_db_instance(**params)
+                changed = True
+            except botocore.exceptions.ClientError as e:
+                e = get_exception()
+                module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:868]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 10]
```diff
@@ -0,0 +1,1148 @@
...
+    if db_instance:
+        response = db_instance
+        changed=False
+    else:
+        try:
+            response = client.restore_db_instance_from_db_snapshot(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:901]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 11]
```diff
@@ -0,0 +1,1148 @@
...
+    # check if instance exists before attempting to reboot
+    db_instance = get_db_instance(client, module)
+    
+    if db_instance:
+        try:
+            response = client.reboot_db_instance(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:928]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 12]
```diff
@@ -0,0 +1,1148 @@
...
+    if db_instance:
+        response = db_instance
+        changed=False
+    else:
+        try:
+            response = client.create_db_instance_read_replica(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:962]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 13]
```diff
@@ -0,0 +1,1148 @@
...
+    if not db_instance:
+        module.fail_json(msg="DB Instance %s does not exist" % instance_name)
+
+    else:
+        try:
+            response = client.promote_read_replica(**params)
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            e = get_exception()
+            module.fail_json(msg=str(e))
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:992]: As comment above - handle as per the guidelines

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 14]
```diff
@@ -0,0 +1,1148 @@
...
+            else:
+                module.fail_json(msg="Parameter %s is not valid for %s command" % (k, command))
+        if k in required_vars:
+            if module.params.get(k):
+                params[v] = module.params[k]
+            else:
+                module.fail_json(msg="Parameter %s required for %s command" % (k, command))
+
+    tag_array = []
+    if module.params.get('tags'):
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:1062]: Use the ec2 helper function `ansible_dict_to_boto3_tag_list`

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 15]
```diff
@@ -0,0 +1,1148 @@
...
+        skip_final_snapshot=dict(type='bool', default=False),
+        final_db_snapshot_identifier=dict(type='str', required=False),
+        force_failover=dict(type='bool', default=False),
+        db_cluster_identifier=dict(type='str', required=False),
+        source_db_instance_identifier=dict(type='str', required=False)
+        )
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:1118]: Define separate checks for json and boto3

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 16]
```diff
@@ -0,0 +1,1148 @@
...
+            'delete': delete_db_instance_or_snapshot,
+            'restore': restore_db_instance_from_snapshot,
+            'reboot': reboot_db_instance,
+            'replicate': replicate_db_instance,
+            'promote': promote_db_instance
+    }        
+
+    changed, results = invocations[module.params.get('command')](client, module)
+
+    module.exit_json(changed=changed, rds=results)
```
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:1141]: Use the ec2 helper function `camel_dict_to_snake_dict` to ensure the results are returned in snake_case format

**Review by wimnat — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/rds_v2.py [hunk 17]
```diff
@@ -0,0 +1,1148 @@
...
+    description:
+      - Database instance identifier. Required except when using command=delete on just a snapshot.
+    required: false
+    default: null
+  engine:
+    description:
+      - The name of the database engine to be used for this instance.  Used only when command=create or restore.
+    required: false
+    default: null
+    choices: [ 'mariadb', 'MySQL', 'oracle-se1', 'oracle-se', 'oracle-ee', 'sqlserver-ee', 'sqlserver-se', 'sqlserver-ex', 'sqlserver-web', 'postgres', 'aurora']
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/rds_v2.py:48]: Is there a way to find this out by introspecting in boto3 instead? Manually updating this is kind of a pain.

**Review by ryansb — COMMENTED** [review]
How hard would it be to be param-compatible with the existing RDS module? It would be preferable not to have a whole second module out for boto3 support. A rewrite to handle boto3 is fine, since the old modules are very boto2-dependent.

**Comment by aioue** [comment]
Closing in favour of original author's prmove. https://github.com/ansible/ansible/pull/19783

**Pull request closed without merging**